src/sentry/eventstream/kafka/backend.py

```diff
@@ -252,14 +254,62 @@ def relay(self, consumer_group, commit_log_topic,
             initial_offset_reset=initial_offset_reset,
         )
 
-        consumer.subscribe([self.publish_topic])
+        owned_partition_offsets = {}
+
+        def on_assign(consumer, partitions):
+            updated_assignment = {(i.topic, i.partition): i.offset for i in partitions}
+
+            # Handle newly assigned partitions by saving them in the owned
+            # partition offsets map.
+            for topic, partition in set(updated_assignment) - set(owned_partition_offsets):
+                logger.debug('Received new partition assignment: %r', (topic, partition))
+
+                offset = updated_assignment[(topic, partition)]
+                if offset is OFFSET_INVALID:
+                    offset = None
+                else:
+                    assert offset > 0, 'unexpected negative offset'
+
+                owned_partition_offsets[(topic, partition)] = offset
+
+            # Handle revoked partitions by removing them from the owned
+            # partition offsets map, and committing  their offsets.
```

Comment:
I have to track down where I verified this, because I have in the past and did for the main Snuba consumer, but I don't think you should do this here.

Going from memory before I track this down: Part of the revoke callback is the broker actually giving the current consumer an opportunity to close up shop. When it returns (or times out...) the broker *then* decides that topic is free to re-assign. By committing offsets for topics in your `on_assign`, you are stomping on another consumers feet (someone else could already be consuming from that partition, because you had no `on_revoke` callback so it was handed off immediately). I guess this is the long round-about answer to your Slack question on why the `on_revoke` callback actually matters.

Reply:
And for other readers, as we discussed in Slack: the current Kafka impl seems to always revoke all assignments on any group change. This made `on_revoke` look a little pointless, but I don't think it's entirely pointless as noted above ^.

We should still do the set/partition tracking that you're doing because we may upgrade Kafka one day and it could start revoking a single partition from your assignments, rather than all of them.

Reply:
I'm struggling to find the proof in documentation. I don't know how Kafka could work if it didn't give owners time to commit offsets *during* the revoke. After the revoke it makes sense to me that all bets are off. Maybe I just tested this with sleeps and stuff?

Reply:
For future readers (me):

Coming from the official client, which is the only well documented client...

https://kafka.apache.org/21/javadoc/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.html

> It is guaranteed that all the processes in a consumer group will execute their onPartitionsRevoked(Collection) callback before any instance executes its onPartitionsAssigned(Collection) callback.